File: python/002. 两数相加.py
"""
@Author: tushushu
@Date: 2018-12-20 15:07:02
"""
# Definition for singly-linked list.


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def get_number(self, l):
        # 将各位数字相加
        head = l
        ret = n = 0
        while head:
            ret += head.val * 10 ** n
            head = head.next
            n += 1
        return ret

    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # 求和
        num = self.get_number(l1) + self.get_number(l2)
        # 结果生成链表
        ret = ListNode(None)
        head = ret
        while 1:
            num, k = divmod(num, 10)
            head.next = ListNode(k)
            head = head.next
            if num == 0:
                break
        return ret.next

File: cpp/002_两数相加.cpp
#include <stddef.h>
using namespace std;

struct ListNode
{
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
  {
    ListNode *ret = new ListNode(0);
    ListNode *head = ret;
    int num, tmp = 0;
    while (l1 != NULL || l2 != NULL)
    {
      num = tmp + (l1 != NULL ? l1->val : 0) + (l2 != NULL ? l2->val : 0);
      tmp = num / 10;
      head->next = new ListNode(num % 10);
      head = head->next;
      if (l1 != NULL)
      {
        l1 = l1->next;
      }
      if (l2 != NULL)
      {
        l2 = l2->next;
      }
    }
    if (tmp)
    {
      head->next = new ListNode(tmp);
    }
    return ret->next;
  }
};